Fix RoboCommand Start producing Task<Task>

RFBomb opened the pull request:
Also fixed PreCompile Directives not targeting targets_OR_GREATER

Resolve issue from this morning's post in #120 

I was running into the issue of starting too many commands with RoboQueue, which was hanging the TaskScheduler implementation since the tasks were running on threads that were being slept, meaning no more tasks could run on those threads. 

SO I had made the sleep and `Process.WaitForExit` asynchonrous.

Well, I found out that `Task.Factory.StartNew` doesn't natively handle Async Tasks, but `Task.Factory.Run` does.
`Task.Factory.Run` silently detects if you are producing `Task<Task>` and unwraps it automatically. `StartNew `is lower level, and you have to unwrap it yourself. I was unaware of this nuance when I made that change. 
After researching the issue, I found a MS article that explains the nuance of `Task.Factory.StartNew ( async() =>` creates a task that returns an async task, so the continuation task fires as soon as the inner async task has done the first 'await'. This caused RoboCommand.Stop() to be called, and reported results that weren't build yet, crashing everything.

I have added `.Unwrap()` to the end of the StartNew() task, which instead of returning Task<Task> will now simply return the inner task. Tested this morning and it works as it used to, but now doesn't block the entire thread since its waiting asynchronously.

RFBomb committed:
Fix RoboCommand Start producing Task<Task>

Also fixed PreCompile Directives not targeting targets_OR_GREATER

PCAssistSoftware commented:
Confirmed that this resolves issue from this morning.

BUT noticed another issue and not sure if related so will await your response before merging

After running a couple of single jobs the overall results 'summary' is empty 

![image](https://user-images.githubusercontent.com/10872799/149661636-67293aa1-62b8-4150-8ee4-8a56b2bef320.png)

RFBomb committed:
Unwrap RoboQueue Task<Task>

RFBomb commented:
I was running into the issue of starting too many commands with RoboQueue, which was hanging the TaskScheduler implementation since the tasks were running on threads that were being slept, meaning no more tasks could run on those threads. 

SO I had made the sleep and `Process.WaitForExit` asynchonrous.

Well, I found out that `Task.Factory.StartNew` doesn't natively handle Async Tasks, but `Task.Factory.Run` does.
`Task.Factory.Run` silently detects if you are producing `Task<Task>` and unwraps it automatically. `StartNew `is lower level, and you have to unwrap it yourself. I was unaware of this nuance when I made that change. 
After researching the issue, I found a MS article that explains the nuance of `Task.Factory.StartNew ( async() =>` creates a task that returns an async task, so the continuation task fires as soon as the inner async task has done the first 'await'. This caused RoboCommand.Stop() to be called, and reported results that weren't build yet, crashing everything.

I have added `.Unwrap()` to the end of the StartNew() task, which instead of returning Task<Task> will now simply return the inner task. Tested this morning and it works as it used to, but now doesn't block the entire thread since its waiting asynchronously.

PCAssistSoftware commented:
@RFBomb - further to my last post above about missing 'overall results - summary

I am right in saying that this now shows differently than before - as 'overall results - summary' now showing source: and destination: which is not relevant for overall - and no longer showing exit code or any jobs cancelled which it did show before?

![image](https://user-images.githubusercontent.com/10872799/149665371-63e14c12-0d46-47da-b747-2b06b98cc967.png)

RFBomb commented:
I have that fixed locall, I ran into a race condition I'm fixing.

PCAssistSoftware commented:
Do you want me to merge this one to fix #120 issue or wait for you to make additional changes?

RFBomb commented:
Additional Changes incoming

RFBomb committed:
Update JobHistoryExpander_SingleJob.xaml.cs

Accomodate the Empty 'PropertyName' in ShowResultsSummary since the EmptyProperty name indicates that all values have changes.

RFBomb committed:
Update ExtensionMethods.cs

Fix the Summary Tag for CancellableSleep

RFBomb committed:
Update RoboQueueProgressEstimator.cs

Proper Async Sleeping

RFBomb committed:
Create Worker Thread for ProgressEstimator

- ProgressEstimator was running into a Race Condition if a RoboCopy was issuing many lines into the log due to many files being in the operation.
- This work was being performed on the main thread that was reacting to the events generated by RoboCommand.Process.
- This resulted in the events generated by Process incrementing the Statistics objects within Progress estimator -> Which in turn triggered updates on the UI.
- This added processing time was creating a race condition where the events were still being issued to resultsBuilder, but the process has already exited and the results were being requested.

SOLUTION:
- Created a worker thread in ProgressEstimator similar to Worker Thread in RoboQueueProgressEstimator. Instead of the `Statistic.PropertyChanged` events occurs immediately as the log lines are pushed to the ResultsBuilder, the lines are reacted to then added to a ConcurrentQueue object that is handled by the CalculationTask.

The CalculationTask pushes updates out every 125ms, and when the ResultsBuilder requestes the results from ProgressEstimator, the cancellation token is triggered.
- This exits the loop within the task that allows it to sleep, and prompts the task to immediately `Task.Delay(250);`. This should be enough time to propogate the remaining events into the queue.  Once the Queues are fully filled out, the results are built.

RFBomb commented:
> I am right in saying that this now shows differently than before - as 'overall results - summary' now showing source: and destination: which is not relevant for overall - and no longer showing exit code or any jobs cancelled which it did show before?

I believe what you are referring to is a quirk of the code. In my opinion its a non-issue. Once a result has been added to the list, it shows properly. Its just that prior to any result being added to the list, it shows the default value.

PCAssistSoftware commented:
That has resolved the issue I had above - are you happy for me to now merge?

PCAssistSoftware commented:
> > I am right in saying that this now shows differently than before - as 'overall results - summary' now showing source: and destination: which is not relevant for overall - and no longer showing exit code or any jobs cancelled which it did show before?
> 
> I believe what you are referring to is a quirk of the code. In my opinion its a non-issue. Once a result has been added to the list, it shows properly. Its just that prior to any result being added to the list, it shows the default value.

Thanks for explanation, agreed - fine with me

RFBomb committed:
Update JobHistoryExpander_SingleJob.xaml.cs

Fix Misleading text on OverallResults Label

RFBomb commented:
> Thanks for explanation, agreed - fine with me
I had a boolean being set in the wrong location in the code. (Set after the string was created instead of before). Resolved.

RFBomb commented:
Ok, this should be mergable now. I think that covers all the things in #120 and based on my testing, its accurately getting the results again. (I had 240 small files being copied on an NVME drive, so the copy operation was finishing before all the log lines were being added to ResultsBuilder since the UI was updating every time a log line was added due to ProgressEstimator's updating of the stats. 

ProgressEstimator now pushes updates every 125ms for the duration of its life, until the internal 'GetResults' method is called.

RFBomb committed:
Update ProgressEstimator.cs

PCAssistSoftware commented:
Just tested here on multiple different sources and all okay, so will merge now
